cluster.go

```diff
@@ -1012,6 +1001,70 @@ func (c *Cluster) Pin(pin api.Pin) error {
 	return err
 }
 
+// validate pin ensures that the metadata accompanying the cid is
+// self-consistent.  This amounts to verifying that the data structure matches
+// the expected form of the pinType carried in the pin.
+func (c *Cluster) validatePin(pin *api.Pin, rplMin, rplMax int) error {
+	switch pin.Type {
+	case api.DataType:
+		if pin.Clusterdag != nil || len(pin.Parents) != 0 {
+			return errors.New("data pins should not reference other pins")
+		}
+	case api.ShardType:
+		if !pin.Recursive {
+			return errors.New("must pin shards recursively")
+		}
+		// In general multiple clusterdags may reference the same shard
+		// and sharder sessions typically update a shard pin's metadata.
+		// Hence we check for an existing shard and carefully update.
+		cState, err := c.consensus.State()
+		if err != nil && err != p2praft.ErrNoState {
+			return err
+		}
+		if err == p2praft.ErrNoState || !cState.Has(pin.Cid) {
+			break
+		}
+
+		// State already tracks pin's CID
+		existing := cState.Get(pin.Cid)
+		// For now all repins of the same shard must use the same
+		// replmax and replmin.  It is unclear what the best UX is here
+		// especially if the same Shard is referenced in multiple
+		// clusterdags.  This simplistic policy avoids complexity and
+		// suits existing needs for shard pins.
+		if existing.ReplicationFactorMin != rplMin ||
+			existing.ReplicationFactorMax != rplMax {
+			return errors.New("shard update with wrong repl factors")
+		}
+		// Pin with union of old and new parents
+		parents := cid.NewSet()
```

Comment:
what I meant is that the `Parents` field in the `Pin` struct should be of type cid.Set. ToSerial() would convert the set to an array (so the serial representation is the same). ToPin() would add the elements in the array to the set. Then you can work directly with the set.

Reply:
Since the serial form doesn't change, you won't need a state migration.

Reply:
I like it!